tensorflow/c/experimental/gradients/model_factory_helper.h

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifndef TENSORFLOW_C_EXPERIMENTAL_GRADIENTS_MODEL_FACTORY_MACRO_H_
+#define TENSORFLOW_C_EXPERIMENTAL_GRADIENTS_MODEL_FACTORY_MACRO_H_
+
+// This macro will expand to a function that defines a `Model`. This `Model` is
```

Comment:
Let's not add these macros right now. We need to fuse the `Model` and `RunModel` impls in `gradients_util.h` and `unified_api_testutil.h` and update the gradient checker to use the non-registry version of `Model` as well.

Reply:
How about a PR for merging `gradients_util.h` and `unified_api_testutil.h`. After that PR is merged, we could rebase it onto this PR which is more cleaner.

Reply:
After some investigations, I think we better do the fusion in a separate PR. `gradient_checker_test` now uses the `Model` defined in `mnist_gradients_testutil.h`. If we want to update the `gradient_checker` to use the non-registry version of `Model`, we have to update `mnist_gradients_testutil.h`( and a lot of files which are dependent on `mnist_gradients_testutil.h` ) and there will be a lot of changes.

I propose we follow this process:
- First, we merge this PR, with the macros.
- Next, I will make a PR for updating these macros to non-registry version along with updating `gradient_checker` and rewriting completely `gradient_checker_test` ( it has overlaps right now ) with the new version of these macros. We can safely remove `mnist_gradients_testutil.h` and `gradient_utils.h` out of `gradient_checker`.
- Next, a PR for moving the tests from `gradients_test` and `mnist_gradients_test` to `module_grad_test`. while moving to `module_grad_test`, we could update the test to use the non-registry version of `Model` and `CompareWithGradientsCheckers`. In addition, we could safely delete `gradients_test` ( not `mnist_gradients_test` because there are some tests for MNIST ).
- Finally, we remove `gradient_utils` and `mnist_gradients_testutil`, update the MNIST tests in `mnist_gradients_test` and maybe move everything related to gradients to `tensorflow/c/experimental/gradients` ( so we could have the same directory structure as python ).

The cleanup process could take a while but I think it's worth doing. What do you think ?

Reply:
I am working on refactoring the `Tape` API that would greatly simplify the model functions. Here's a draft [proposal](https://github.com/tensorflow/community/compare/master...saxenasaurabh:cpp-gradients?expand=1&short_path=c3fb7c7#diff-c3fb7c75a838948b459de9b64e6edc8d2f75b49fdf7dad8040f6724f00fb85f8) that I will publish as a formal RFC in a day or so (let me know if you have any suggestions!). I have the code mostly ready for that. Once that is done the macro would look like:

```
#define TF_GRAD_MODEL_FACTORY(name, num_inputs, num_outputs, num_grad_outputs, \
                              ops_call)                                        \
  Status name(AbstractContext* ctx,                                            \
              absl::Span<AbstractTensorHandle* const> inputs,                  \
              absl::Span<AbstractTensorHandle*> outputs,                       \
              const GradientRegistry& registry) {                              \
    std::unique_ptr<Tape>(new Tape(/*persistent=*/false));                                \
    for (int i{}; i < num_inputs; ++i) {                                       \
      tape->Watch(inputs[i]);                                            \
    }                                                                          \
                                                                               \
    AbstractTensorHandle* temp_outputs[num_outputs] = {};                      \
    AbstractContextPtr tape_ctx(new TapeContext(ctx, tape, registry));         \
    ops_call;                                                                  \
    TF_RETURN_IF_ERROR(tape->ComputeGradient(                                  \
        ctx, temp_outputs, inputs,                          \
        /*output_gradients=*/{}, outputs));                                 \
                                                                               \
    for (int i{}; i < num_outputs; ++i) {                                      \
      temp_outputs[i]->Unref();                                                \
    }                                                                          \
    return Status::OK();                                                       \
  }
```

If we want we could further add helpers to watch a list of tensors. Also we need to figure out how to avoid manually `Unref`ing intermediate tensors. I have an idea for this that might require a larger refactor. Once all this is done the model fn will be a fairly short. 

Another problem with this macro style is that you rely on variables created in the macro in your `ops_call`. That does not seem like a great pattern to me.

My suggestion for now would be to drop the macros and use the verbose model function. Then think about how to de-dup the `Model` and `RunModel` and move existing tests to `module_grad_test`. That will be a great cleanup I think. Meanwhile I will try to check in the new API. Once that is done we can discuss what additional helpers we need to make the tests less verbose.

Reply:
Done